<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="deconverse">
<title>Introduction: full explained deconverse pipeline with PBMCs • deconverse</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction: full explained deconverse pipeline with PBMCs">
<meta property="og:description" content="deconverse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">deconverse</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.6.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/intro_pbmc.html">Introduction: full explained deconverse pipeline with PBMCs</a>
    <a class="dropdown-item" href="../articles/pdac_tutorial.html">Benchmarking example with pancreatic cancer data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction: full explained deconverse pipeline with PBMCs</h1>
                        <h4 data-toc-skip class="author">Clarice S.
Groeneveld</h4>
            
      
      
      <div class="d-none name"><code>intro_pbmc.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="context">Context<a class="anchor" aria-label="anchor" href="#context"></a>
</h2>
<p>We’ll demonstrate the functionality of <code>deconverse</code> using
a PBMC dataset. This is a small single-cell reference with a limited
number of cells per cell type particularly for the lower levels of cell
type hierarchy.</p>
<p><code>deconverse</code> implements multiple bulk deconvolution
methods based on single-cell references, and enhances them by adding
hierarchy-awareness. It does so by deconvoluting at the different levels
of the cell type hierarchy and using results from the higher level of
hierarchy to correct the lower level estimated fractions.</p>
<p>In this example, that comes provided with the <code>Seurat</code>
package, peripheral blood mononuclear cells (PBMCs) are annotated at 2
levels of hierarchy. Bulk RNA-seq data can be deconvoluted at both
levels.</p>
<p>Usually, we would recommend building the single-cell reference on one
dataset and benchmarking on a different one. However, for this example,
we will split cells into training (reference) and test (benchmarking)
set. Benchmarking is not a necessary step of <code>deconverse</code>,
but a feature that can be used to compare the different implemented
methods in your data type and decide what levels of annotation are
trustworthy.</p>
</div>
<div class="section level2">
<h2 id="load-pbmc-dataset">Load PBMC dataset<a class="anchor" aria-label="anchor" href="#load-pbmc-dataset"></a>
</h2>
<p>First, we load the dataset from Seurat and look at the provided data
and its two levels of annotation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">deconverse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13714 features across 2624 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13714 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Cell_major_identities"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"Cell_minor_identities"</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,  label.size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>          axis.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>          plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-4-1.png" width="624"></p>
<p>PBMCs are annotated at two levels, encoded in the variables
<code>Cell_major_identities</code> and
<code>Cell_minor_identities</code> in the object metadata. The UMAP
divides the cells into three major clusters, corresponding to the
<code>Cell_major_identities</code> which we will refer to as the first
level of annotation or <code>l1</code>. This is broken down into
finer-grained annotation in the second level of annotation or
<code>l2</code>.</p>
</div>
<div class="section level2">
<h2 id="split-traintest">Split train/test<a class="anchor" aria-label="anchor" href="#split-traintest"></a>
</h2>
<p>To run both the reference and the benchmarking from this data, we
need to split the cells randomly into training (60%) and test (40%)
sets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">ncells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">train_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ncells</span>, <span class="va">ncells</span><span class="op">*</span><span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="va">test_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ncells</span>, <span class="va">train_ids</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">pbmc_train</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span>,<span class="va">train_ids</span><span class="op">]</span></span>
<span><span class="va">pbmc_test</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span>,<span class="va">test_ids</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_train</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Cell_major_identities"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_test</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Cell_major_identities"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>          axis.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>          plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-6-1.png" width="624"></p>
</div>
<div class="section level2">
<h2 id="screference-build-single-cell-reference-from-train-data">
<code>screference</code>: Build single-cell reference from train
data<a class="anchor" aria-label="anchor" href="#screference-build-single-cell-reference-from-train-data"></a>
</h2>
<p>Now, we’re ready to build the reference as an
<code>hscreference</code> object. We use
<code><a href="../reference/new_screference.html">deconverse::new_screference</a></code> and give our Seurat object
<code>pbmc_train</code>, the l1 and l2 annotation variables in order, a
project name, and a batch_id variable which could correspond to patient
in another context, but here is just the same value for all cells.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/new_screference.html">new_screference</a></span><span class="op">(</span><span class="va">pbmc_train</span>,</span>
<span>                annot_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell_major_identities"</span>, <span class="st">"Cell_minor_identities"</span><span class="op">)</span>,</span>
<span>                project_name <span class="op">=</span> <span class="st">"pbmc_example"</span>,</span>
<span>                batch_id <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating new `hscreference` object...</span></span>
<span><span class="va">pbmc_ref</span></span>
<span><span class="co">#&gt; h-screference object named `pbmc_example` with 2 levels of annotation from 1574 cells</span></span>
<span><span class="co">#&gt; population annotation tree: </span></span>
<span><span class="co">#&gt;                levelName</span></span>
<span><span class="co">#&gt; 1  populations          </span></span>
<span><span class="co">#&gt; 2   ¦--TNK              </span></span>
<span><span class="co">#&gt; 3   ¦   ¦--Naive CD4 T  </span></span>
<span><span class="co">#&gt; 4   ¦   ¦--Memory CD4 T </span></span>
<span><span class="co">#&gt; 5   ¦   ¦--CD8 T        </span></span>
<span><span class="co">#&gt; 6   ¦   °--NK           </span></span>
<span><span class="co">#&gt; 7   ¦--B                </span></span>
<span><span class="co">#&gt; 8   ¦   °--B            </span></span>
<span><span class="co">#&gt; 9   °--Monocytic_lineage</span></span>
<span><span class="co">#&gt; 10      ¦--CD14+ Mono   </span></span>
<span><span class="co">#&gt; 11      ¦--FCGR3A+ Mono </span></span>
<span><span class="co">#&gt; 12      °--DC           </span></span>
<span><span class="co">#&gt; cached results: </span></span>
<span><span class="co">#&gt; l1:  </span></span>
<span><span class="co">#&gt; l2:</span></span></code></pre></div>
<p>Once the object is created, it can be used to compute a reference for
different methods. We’ll compute references for DWLS (which is also used
for OLS and SVR), svr and AutoGeneS. For all available deconvolution
methods, use <code>deconvolution_methods()</code>.</p>
<p>And then compute the references (they will be grabbed from the
cache):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Optional: parallelize with Seurat</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize<span class="op">=</span> <span class="fl">4000</span><span class="op">*</span><span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multicore"</span>, workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in supportsMulticoreAndRStudio(...): [ONE-TIME WARNING] Forked</span></span>
<span><span class="co">#&gt; processing ('multicore') is not supported when running R from RStudio because</span></span>
<span><span class="co">#&gt; it is considered unstable. For more details, how to control forked processing</span></span>
<span><span class="co">#&gt; or not, and how to silence this warning in future R sessions, see</span></span>
<span><span class="co">#&gt; ?parallelly::supportsMulticore</span></span>
<span></span>
<span><span class="co"># Compute references</span></span>
<span><span class="va">pbmc_ref</span> <span class="op">&lt;-</span> <span class="va">pbmc_ref</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">compute_reference</span><span class="op">(</span><span class="st">"dwls"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">compute_reference</span><span class="op">(</span><span class="st">"autogenes"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Results found in cache, returning...</span></span>
<span><span class="co">#&gt; Results found in cache, returning...</span></span>
<span><span class="co">#&gt; Results found in cache, returning...</span></span>
<span><span class="co">#&gt; Results found in cache, returning...</span></span></code></pre></div>
<p>Now the single-cell reference is ready for deconvolution! First,
let’s benchmark it with our test set.</p>
<div class="section level3">
<h3 id="computational-performance">Computational performance<a class="anchor" aria-label="anchor" href="#computational-performance"></a>
</h3>
<p>We can plot computational performance to compare the methods. This
information is also available as a table
(<code>pbmc_ref$metrics</code>)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plt_comp_performance</span><span class="op">(</span><span class="va">pbmc_ref</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-10-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="apply-to-real-pbmc-bulk-rna-seq">Apply to real PBMC bulk RNA-seq<a class="anchor" aria-label="anchor" href="#apply-to-real-pbmc-bulk-rna-seq"></a>
</h2>
<p>To demonstrate the real-life usability of deconverse, we will use an
example dataset of bulk PBMCs.</p>
<div class="section level3">
<h3 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h3>
<p>First, we download the data and keep only the PBMC bulk samples.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/seandavi/GEOquery" class="external-link">GEOquery</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Setting options('download.file.method.GEOquery'='auto')</span></span>
<span><span class="co">#&gt; Setting options('GEOquery.inmemory.gpl'=FALSE)</span></span>
<span><span class="co"># Get dataset</span></span>
<span><span class="va">gset</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://seandavi.github.io/GEOquery/reference/getGEO.html" class="external-link">getGEO</a></span><span class="op">(</span><span class="st">"GSE107011"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Found 1 file(s)</span></span>
<span><span class="co">#&gt; GSE107011_series_matrix.txt.gz</span></span>
<span><span class="co"># Get sample annotation</span></span>
<span><span class="va">samp_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">gset</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tibble</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">source_name_ch1</span> <span class="op">==</span> <span class="st">"PBMCs"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">title</span>, <span class="st">"_rep.*$"</span><span class="op">)</span>,</span>
<span>           sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu">str_detect</span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"^\\d"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="va">sample_name</span><span class="op">)</span>, <span class="va">sample_name</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Download gene expression matrix</span></span>
<span><span class="fu">curl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/curl/man/curl_download.html" class="external-link">curl_download</a></span><span class="op">(</span><span class="st">"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE107nnn/GSE107011/suppl/GSE107011_Processed_data_TPM.txt.gz"</span>, <span class="st">"GSE107011_TPM.txt.gz"</span><span class="op">)</span></span>
<span><span class="va">gexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"GSE107011_TPM.txt.gz"</span><span class="op">)</span></span>
<span><span class="va">gexp</span> <span class="op">&lt;-</span> <span class="va">gexp</span><span class="op">[</span>,<span class="va">samp_annot</span><span class="op">$</span><span class="va">sample_name</span><span class="op">]</span></span>
<span><span class="co"># Remove gene ID version</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">str_remove</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span>, <span class="st">"\\.[0-9]+$"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"GSE107011_TPM.txt.gz"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-gene-annotation">Get gene annotation<a class="anchor" aria-label="anchor" href="#get-gene-annotation"></a>
</h3>
<p>As gene annotation was not provided, we download and add annotation
used to pre-process the data according to their documentation to get
gene symbols. Ideally, annotation should be from the same version of
Ensembl, but to simplify, we’ll use the ‘latest’.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/biomaRt" class="external-link">biomaRt</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshkatz/needs" class="external-link">needs</a></span><span class="op">)</span></span>
<span><span class="fu">needs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/needs/man/prioritize.html" class="external-link">prioritize</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span>
<span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html" class="external-link">useEnsembl</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"genes"</span>, </span>
<span>                      dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html" class="external-link">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>, <span class="st">"hgnc_symbol"</span><span class="op">)</span>,</span>
<span>                    filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span>,</span>
<span>                    <span class="va">ensembl</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="va">gene_annot</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">hgnc_symbol</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hgnc_symbol</span> <span class="op">!=</span> <span class="st">""</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">hgnc_symbol</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use gene symbols (like in reference)</span></span>
<span><span class="va">gexp</span> <span class="op">&lt;-</span> <span class="va">gexp</span><span class="op">[</span><span class="va">gene_annot</span><span class="op">$</span><span class="va">ensembl_gene_id</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_annot</span><span class="op">$</span><span class="va">hgnc_symbol</span></span>
<span><span class="va">gexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 40682    13</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-deconverse">Run deconverse<a class="anchor" aria-label="anchor" href="#run-deconverse"></a>
</h3>
<p>Now we can run <code>deconverse</code> on the bulk PBMC dataset with
symbol identifiers from the reference we previously calculated.</p>
<p><strong>Note:</strong> <strong>the <code>pbmc_bench</code> is not
necessary for any of these steps</strong>, it only serves to benchmark
the different methods.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconv_res</span> <span class="op">&lt;-</span> <span class="fu">deconvolute_all</span><span class="op">(</span><span class="va">gexp</span>, <span class="va">pbmc_ref</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dwls"</span>, <span class="st">"ols"</span>, <span class="st">"svr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compare-results">Compare results<a class="anchor" aria-label="anchor" href="#compare-results"></a>
</h3>
<p>Now we cal compare the results at the two levels of hierarchy.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l1_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">deconv_res</span>, <span class="st">"[["</span>, <span class="st">"l1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">l2_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">deconv_res</span>, <span class="st">"[["</span>, <span class="st">"l2"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">frac_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">l1_results</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">str_subset</span><span class="op">(</span><span class="st">"^frac"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_frac_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">frac_names</span>, <span class="kw">function</span><span class="op">(</span><span class="va">frac</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">frac</span>, <span class="st">"^frac_"</span><span class="op">)</span></span>
<span>    <span class="va">l1_results</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">pivot_wider</span><span class="op">(</span>id_cols <span class="op">=</span> <span class="st">"sample"</span>, names_from <span class="op">=</span> <span class="st">"method"</span>, values_from <span class="op">=</span> <span class="va">frac</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_"</span>, <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># install.packages("ggcorrplot")</span></span>
<span><span class="fu">ggcorrplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html" class="external-link">ggcorrplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">all_frac_mat</span><span class="op">)</span>, hc.order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-15-1.png" width="384"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">frac_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">l2_results</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">str_subset</span><span class="op">(</span><span class="st">"^frac"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_frac_mat_l2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">frac_names</span>, <span class="kw">function</span><span class="op">(</span><span class="va">frac</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">frac</span>, <span class="st">"^frac_"</span><span class="op">)</span></span>
<span>    <span class="va">l2_results</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu">pivot_wider</span><span class="op">(</span>id_cols <span class="op">=</span> <span class="st">"sample"</span>, names_from <span class="op">=</span> <span class="st">"method"</span>, values_from <span class="op">=</span> <span class="va">frac</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_"</span>, <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggcorrplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html" class="external-link">ggcorrplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">all_frac_mat_l2</span>, use <span class="op">=</span> <span class="st">"pair"</span><span class="op">)</span>, hc.order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-16-1.png" width="384"></p>
<p>We see that there is a much higher correlation between methods at the
coarse-level of annotation, as expected.</p>
</div>
<div class="section level3">
<h3 id="check-sample-deconvolution-by-a-method">Check sample deconvolution by a method<a class="anchor" aria-label="anchor" href="#check-sample-deconvolution-by-a-method"></a>
</h3>
<p>We can verify the results of deconvolution by plotting the
proportions:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">pivot_dwls_l1</span> <span class="op">&lt;-</span> <span class="va">l1_results</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"DWLS"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"frac"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"pop"</span>, values_to <span class="op">=</span> <span class="st">"frac"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>pop <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">pop</span>, <span class="st">"frac_"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"Monocytic_lineage"</span>, <span class="st">"TNK"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           level <span class="op">=</span> <span class="st">"l1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pivot_dwls_l2</span> <span class="op">&lt;-</span> <span class="va">l2_results</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"DWLS"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"frac"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"pop"</span>, values_to <span class="op">=</span> <span class="st">"frac"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>pop <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">pop</span>, <span class="st">"frac_"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"DC"</span>,</span>
<span>                                                                <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"CD8 T"</span>,</span>
<span>                                                                <span class="st">"Memory CD4 T"</span>, <span class="st">"Naive CD4 T"</span>,</span>
<span>                                                                <span class="st">"NK"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           level <span class="op">=</span> <span class="st">"l2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pop_order</span> <span class="op">&lt;-</span> <span class="va">pivot_dwls_l1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op">==</span> <span class="st">"TNK"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">arrange</span><span class="op">(</span><span class="va">frac</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="va">l1_plt</span> <span class="op">&lt;-</span> <span class="va">pivot_dwls_l1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sample</span>, levels <span class="op">=</span> <span class="va">pop_order</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">sample</span>, <span class="va">frac</span>, fill <span class="op">=</span> <span class="va">pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_col</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span><span class="op">)</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggheatmapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggheatmapper/man/theme_sparse.html" class="external-link">theme_scatter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"l1"</span>, y <span class="op">=</span> <span class="st">"fraction"</span>, fill <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="va">l2_plt</span> <span class="op">&lt;-</span> <span class="va">pivot_dwls_l2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sample</span>, levels <span class="op">=</span> <span class="va">pop_order</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">sample</span>, <span class="va">frac</span>, fill <span class="op">=</span> <span class="va">pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_col</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span><span class="op">)</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggheatmapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggheatmapper/man/theme_sparse.html" class="external-link">theme_scatter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"l2"</span>, y <span class="op">=</span> <span class="st">"fraction"</span>, fill <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">l1_plt</span> <span class="op">/</span> <span class="va">l2_plt</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-17-1.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="scbench-run-benchmarking-of-methods-on-bespoke-data">
<code>scbench</code>: Run benchmarking of methods on bespoke
data<a class="anchor" aria-label="anchor" href="#scbench-run-benchmarking-of-methods-on-bespoke-data"></a>
</h2>
<div class="section level3">
<h3 id="bounds">Bounds<a class="anchor" aria-label="anchor" href="#bounds"></a>
</h3>
<p>Now, let’s build our benchmarking set. First, we need to understand
how the annotations relate to each other to build the “bounds” of our
pseudobulk simulations (don’t worry, this will be further
explained).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pbmc_test</span><span class="op">$</span><span class="va">Cell_major_identities</span>, <span class="va">pbmc_test</span><span class="op">$</span><span class="va">Cell_minor_identities</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    </span></span>
<span><span class="co">#&gt;                     Naive CD4 T CD14+ Mono Memory CD4 T   B CD8 T FCGR3A+ Mono</span></span>
<span><span class="co">#&gt;   TNK                       269          0          202   0   117            0</span></span>
<span><span class="co">#&gt;   B                           0          0            0 128     0            0</span></span>
<span><span class="co">#&gt;   Monocytic_lineage           0        187            0   0     0           69</span></span>
<span><span class="co">#&gt;                    </span></span>
<span><span class="co">#&gt;                      NK  DC</span></span>
<span><span class="co">#&gt;   TNK                64   0</span></span>
<span><span class="co">#&gt;   B                   0   0</span></span>
<span><span class="co">#&gt;   Monocytic_lineage   0  14</span></span></code></pre></div>
<p>We can see that <code>l1 B</code> cells only correspond to
<code>l2 B</code> cells; <code>l1 Monocytic_lineage</code> becomes
<code>l2 CD14+ Mono</code>, <code>FCGR3A+ Mono</code> and
<code>DC</code>; and <code>l1 TNK</code> splits into
<code>l2 CD8 T</code>, <code>Memory CD4 T</code>,
<code>Naive CD4 T</code> and <code>NK</code>.</p>
<p>In PBMCs, lymphocytes make up about 70-90% of cells and monocytes
make up 10-20%. If we want to make pseudobulk mixtures that resemble
real PBMCs, we should include in them more lymphocytes, particularly
more abundant T cells, than monocytes. This is what bounds are for.</p>
<p>Bounds are used to create random mixtures with controlled
proportions. For each population, we need a lower and upper bound. We’ll
start with <code>l1</code> populations:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l1_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"Monocytic_lineage"</span>, <span class="st">"TNK"</span><span class="op">)</span>,</span>
<span>    lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.85</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We’ve set that B cells will make up between 10-20% of mixtures,
Monocytic lineage between 10-30% and TNK 50-85%. The structure of bounds
should always be a <code>data.frame</code> with 3 columns (population,
lower, upper). Make sure the population names are the same as the levels
in the annotation table.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pbmc_test</span><span class="op">$</span><span class="va">Cell_major_identities</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] B                 Monocytic_lineage TNK              </span></span>
<span><span class="co">#&gt; Levels: TNK B Monocytic_lineage</span></span></code></pre></div>
<p>Bounds at finer levels of l2 are always given within the context of
their coarser grained annotation. For example,
<code>Monocytic lineage</code> will be split into 3 populations. As
dendritic cells are more rare than monocytes, we can make them more rare
in our mixtures as well:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l2_mono_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14+ Mono"</span>, <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"DC"</span><span class="op">)</span>,</span>
<span>        lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.6</span>, <span class="fl">0.2</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>Following the same logic, we can do the bounds for the
sub-populations of <code>TNK</code>. It’s not necessary for
<code>B</code>, since they are not split into sub-populations in this
annotation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l2_tnk_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8 T"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"Naive CD4 T"</span>, <span class="st">"NK"</span><span class="op">)</span>,</span>
<span>        lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>        upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.3</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>All bounds should be put together into a list, and names should be l1
for the coarsest-grained population and
<code>{level}_{coarse_population_that_was_split}</code> for finer levels
(e.g. <code>l2_Monocytic_lineage</code> ).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    l1 <span class="op">=</span> <span class="va">l1_bounds</span>,</span>
<span>    l2_TNK <span class="op">=</span> <span class="va">l2_tnk_bounds</span>,</span>
<span>    l2_Monocytic_lineage <span class="op">=</span> <span class="va">l2_mono_bounds</span> </span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-the-scbench">Create the <code>scbench</code><a class="anchor" aria-label="anchor" href="#create-the-scbench"></a>
</h3>
<p>Now we have the information we need to create our benchmarking set.
Variables are very similar to creating an <code>screference</code>, with
the exception of <code>pop_bounds</code>, which we have explained
above.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_bench</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/new_scbench.html">new_scbench</a></span><span class="op">(</span><span class="va">pbmc_test</span>, </span>
<span>                         annot_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell_major_identities"</span>,</span>
<span>                                       <span class="st">"Cell_minor_identities"</span><span class="op">)</span>,</span>
<span>                         pop_bounds <span class="op">=</span> <span class="va">pbmc_bounds</span>,</span>
<span>                         project_name <span class="op">=</span> <span class="st">"pbmc_example"</span>,</span>
<span>                         batch_id <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating new `scbench` object...</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(l1)`</span></span>
<span><span class="va">pbmc_bench</span></span>
<span><span class="co">#&gt; scbench object named pbmc_example with 11 reference populations and 2 levels of annotation</span></span>
<span><span class="co">#&gt; Bounds [x] | Mixtures [ ] | Spillover [ ] | Limit of Detection [ ] | Pseudobulks [ ]</span></span></code></pre></div>
<p>There are 3 main benchmarks tested by scbench:</p>
<ul>
<li><p>population</p></li>
<li><p>limit of detection (lod)</p></li>
<li><p>spillover</p></li>
</ul>
<p>The <strong>population</strong> benchmark, pseudobulk mixtures are
created using the random proportions constrained by the bounds. These
known proportions will be then compared with the fractions estimated by
the deconvolution methods to assess their accuracy.</p>
<p>The <strong>limit of detection</strong> benchmark estimates how low
the fraction of a particular population has to be for a method to be
able to detect it.</p>
<p>In the <strong>spillover</strong> benchmark, cells for different
pairs of populations on the same level are mixed with increasing
proportions to evaluate whether the methods can distinguish them well
from each other.</p>
<p>The first step is to generate the mixtures for the 3 different
methods:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_bench</span> <span class="op">&lt;-</span> <span class="va">pbmc_bench</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/mixtures_population.html">mixtures_population</a></span><span class="op">(</span>nsamp <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/mixtures_lod.html">mixtures_lod</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/mixtures_spillover.html">mixtures_spillover</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simulating spillover mixtures between population pairs...</span></span>
<span><span class="co">#&gt; Simulating limits of detection for each population...</span></span>
<span><span class="co">#&gt; Simulating population mixtures given bounds...</span></span></code></pre></div>
<p>Then, we can generate pseudobulks for these mixtures by mixing 100
random cells that correspond to the proportions given by the mixtures.
Finally, we can call the different methods to deconvolute.</p>
<p>As for the screference, the next steps will go a lot faster by using
cached results:</p>
<p>Here, we’re demonstrating with 6 methods currently implemented. For a
full list, see <code>deconvolution_methods()</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_bench</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pseudobulks.html">pseudobulks</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, ncells <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Pseudobulks for population found in cache. Skipping...</span></span>
<span><span class="co">#&gt; Pseudobulks for lod found in cache. Skipping...</span></span>
<span><span class="co">#&gt; Pseudobulks for lod found in cache. Skipping...</span></span>
<span><span class="co">#&gt; Pseudobulks for spillover found in cache. Skipping...</span></span>
<span><span class="co">#&gt; Pseudobulks for spillover found in cache. Skipping...</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_bench</span> <span class="op">&lt;-</span> <span class="fu">deconvolute_all</span><span class="op">(</span><span class="va">pbmc_bench</span>, <span class="va">pbmc_ref</span>,</span>
<span>                              methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dwls"</span>, <span class="st">"svr"</span>, <span class="st">"ols"</span>, <span class="st">"autogenes"</span>, <span class="st">"bisque"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="benchmarking-results">Benchmarking results<a class="anchor" aria-label="anchor" href="#benchmarking-results"></a>
</h3>
<p>deconverse implements many functions to produce many benchmarking
plots, either for a single method or summarizing and comparing multiple
methods.</p>
<div class="section level4">
<h4 id="evaluation-of-a-single-method">Evaluation of a single method<a class="anchor" aria-label="anchor" href="#evaluation-of-a-single-method"></a>
</h4>
<p>Here, we generate all evaluation plots for results coming from the
dampened weighted least squares (<code>dwls</code>).
<code>plt_cors_scatter</code> plots the correlations between
deconvoluted pseudobulk portions found by the method and the true
fractions.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_cors_scatter.html">plt_cors_scatter</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-29-1.png" width="576"></p>
<p><code>plt_lod_scatter</code> plots correlations between pseudobulk
mixtures with increasing proportion of one population and where the
limit of detection was established for that method.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_lod_scatter.html">plt_lod_scatter</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-30-1.png" width="576"></p>
<p><code>plt_spillover_scatter</code> plots the spillover between each
pair of populations to help identify pairs of populations that will
possibly be confused by the two methods.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_spillover_scatter.html">plt_spillover_scatter</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-31-1.png" width="624"></p>
<p>By default, plots will show the benchmark on finest grained
annotation available. We can verify on coarser grained by specifying the
level:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_cors_scatter.html">plt_cors_scatter</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, method <span class="op">=</span> <span class="st">"dwls"</span>, level <span class="op">=</span> <span class="st">"l1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-32-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="comparison-of-methods">Comparison of methods<a class="anchor" aria-label="anchor" href="#comparison-of-methods"></a>
</h4>
<p>There are also plots available to compare performance on the same
data from different methods. The two main heatmaps compare correlation
between pseudobulk mixture proportions and deconvoluted fractions by
population, or the root mean squared error (RMSE) from predicted
fraction and real fraction.</p>
<div class="section level5">
<h5 id="population">Population<a class="anchor" aria-label="anchor" href="#population"></a>
</h5>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_cor_heatmap.html">plt_cor_heatmap</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, level <span class="op">=</span> <span class="st">"l2"</span><span class="op">)</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-33-1.png" width="576"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_rmse_heatmap.html">plt_rmse_heatmap</a></span><span class="op">(</span><span class="va">pbmc_bench</span>, level <span class="op">=</span> <span class="st">"l2"</span><span class="op">)</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-34-1.png" width="576"></p>
</div>
<div class="section level5">
<h5 id="lod-and-spillover">LoD and Spillover<a class="anchor" aria-label="anchor" href="#lod-and-spillover"></a>
</h5>
<p>Plots are also available for limit of detection (lower is better) and
spillover RMSE (lower is better).</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_lod_heatmap.html">plt_lod_heatmap</a></span><span class="op">(</span><span class="va">pbmc_bench</span><span class="op">)</span><span class="op">$</span><span class="va">heatmap</span></span>
<span><span class="co">#&gt; Warning: Removed 2 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_text()`).</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-35-1.png" width="576"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plt_spillover_heatmap.html">plt_spillover_heatmap</a></span><span class="op">(</span><span class="va">pbmc_bench</span><span class="op">)</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-36-1.png" width="576"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="computational-performance-1">Computational performance<a class="anchor" aria-label="anchor" href="#computational-performance-1"></a>
</h3>
<p>We can also plot computational performance metrics easily for
<code>scbench</code> in the same way we did for <code>scref</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plt_comp_performance</span><span class="op">(</span><span class="va">pbmc_bench</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_pbmc_files/figure-html/unnamed-chunk-37-1.png" width="576">
DWLS is the slowest method of this comparison even if we have
pre-computed the reference. Bisque is more memory hungry, but it
computes the reference in the <code>deconvolute</code> step.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Clarice S Groeneveld.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
