<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ deconverse</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://csgroen.github.io/deconverse/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">deconverse</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.6.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/intro_pbmc.html">Introduction: full explained deconverse pipeline with PBMCs</a>
    <a class="dropdown-item" href="../articles/pdac_tutorial.html">Benchmarking example with pancreatic cancer data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a></h2>
      
      <p class="section-desc"></p><p>Functions to build and interact with (hierarchical) single-cell reference objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="new_screference.html">new_screference()</a></code> 
        </dt>
        <dd>Generate single-cell reference (`screference`) object</dd>
      </dl><dl><dt>
          
          <code><a href="new_hscreference.html">new_hscreference()</a></code> 
        </dt>
        <dd>Generate multi-level hierarchical single-cell reference (`hscreference`) object</dd>
      </dl><dl><dt>
          
          <code><a href="compute_reference.html">compute_reference(<i>&lt;screference&gt;</i>)</a></code> <code><a href="compute_reference.html">compute_reference(<i>&lt;hscreference&gt;</i>)</a></code> 
        </dt>
        <dd>Compute reference given a deconvolution method from an `screference` object</dd>
      </dl><dl><dt>
          
          <code><a href="autogenes_scref.html">autogenes_scref()</a></code> 
        </dt>
        <dd>Compute reference matrix from an `screference` object using
AutoGeneS</dd>
      </dl><dl><dt>
          
          <code><a href="bayesprism_scref.html">bayesprism_scref()</a></code> 
        </dt>
        <dd>Compute reference matrix from an `screference` object using
BayesPrism</dd>
      </dl><dl><dt>
          
          <code><a href="cibersortx_scref.html">cibersortx_scref()</a></code> 
        </dt>
        <dd>Compute reference signature matrix from `screference` object using
method implemented in CIBERSORTx</dd>
      </dl><dl><dt>
          
          <code><a href="dwls_scref.html">dwls_scref()</a></code> 
        </dt>
        <dd>Compute reference signature matrix from `screference` object using
method implemented in DWLS package for methods `dwls`, `svr` and `ols`</dd>
      </dl><dl><dt>
          
          <code><a href="scaden_scref.html">scaden_scref()</a></code> 
        </dt>
        <dd>Compute reference matrix from an `screference` object using scaden (Single-cell assisted deconvolutional network)</dd>
      </dl></div><div class="section level2">
      <h2 id="benchmarking">Benchmarking<a class="anchor" aria-label="anchor" href="#benchmarking"></a></h2>
      
      <p class="section-desc"></p><p>Functions to create a benchmarking scbench object.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="new_scbench.html">new_scbench()</a></code> 
        </dt>
        <dd>Generate `scbench` object</dd>
      </dl><dl><dt>
          
          <code><a href="mixtures_lod.html">mixtures_lod()</a></code> 
        </dt>
        <dd>Simulate mixtures to find the limit of detection of each population in
mixtures in an `scbench` object</dd>
      </dl><dl><dt>
          
          <code><a href="mixtures_population.html">mixtures_population()</a></code> 
        </dt>
        <dd>Simulate sample population mixtures given bounds in an `scbench` object</dd>
      </dl><dl><dt>
          
          <code><a href="mixtures_spillover.html">mixtures_spillover()</a></code> 
        </dt>
        <dd>Simulate mixtures for measuring spillover between all pairs of populations
in an `scbench` object</dd>
      </dl><dl><dt>
          
          <code><a href="pseudobulks.html">pseudobulks()</a></code> 
        </dt>
        <dd>Generate pseudobulk gene expression from single-cell RNA-seq given
population mixtures in an `scbench` object</dd>
      </dl></div><div class="section level2">
      <h2 id="deconvolution">Deconvolution<a class="anchor" aria-label="anchor" href="#deconvolution"></a></h2>
      
      <p class="section-desc"></p><p>Functions to run deconvolution based on a single-cell reference</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="deconvolute.html">deconvolute(<i>&lt;scbench&gt;</i>)</a></code> <code><a href="deconvolute.html">deconvolute(<i>&lt;matrix&gt;</i>)</a></code> 
        </dt>
        <dd>Deconvolute `scbench` object using a chosen method</dd>
      </dl><dl><dt>
          
          <code><a href="deconvolute_all.html">deconvolute_all(<i>&lt;scbench&gt;</i>)</a></code> <code><a href="deconvolute_all.html">deconvolute_all(<i>&lt;matrix&gt;</i>)</a></code> 
        </dt>
        <dd>Deconvolute `scbench` object using all methods with default settings</dd>
      </dl><dl><dt>
          
          <code><a href="autogenes_deconvolute.html">autogenes_deconvolute()</a></code> 
        </dt>
        <dd>AutoGeneS deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="bayesprism_deconvolute.html">bayesprism_deconvolute()</a></code> 
        </dt>
        <dd>BayesPrism deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="bisque_deconvolute.html">bisque_deconvolute()</a></code> 
        </dt>
        <dd>Bisque deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="cibersortx_deconvolute.html">cibersortx_deconvolute()</a></code> 
        </dt>
        <dd>CIBERSORTx deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="dwls_deconvolute.html">dwls_deconvolute()</a></code> 
        </dt>
        <dd>Dampened weighted least squares (DWLS) deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="music_deconvolute.html">music_deconvolute()</a></code> 
        </dt>
        <dd>MuSiC deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="ols_deconvolute.html">ols_deconvolute()</a></code> 
        </dt>
        <dd>Ordinary least squares (OLS) deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="scaden_deconvolute.html">scaden_deconvolute()</a></code> 
        </dt>
        <dd>scaden (Single-cell assisted deconvolutional network) deconvolution of bulk data using an `screference`</dd>
      </dl><dl><dt>
          
          <code><a href="svr_deconvolute.html">svr_deconvolute()</a></code> 
        </dt>
        <dd>Support vector regression (SVR) deconvolution of bulk data using an `screference`</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting-and-results">Plotting and Results<a class="anchor" aria-label="anchor" href="#plotting-and-results"></a></h2>
      
      <p class="section-desc"></p><p>Functions to recover benchmarking results</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="results.scbench.html">results(<i>&lt;scbench&gt;</i>)</a></code> 
        </dt>
        <dd>Get summary of benchmark results for an analysis type</dd>
      </dl><dl><dt>
          
          <code><a href="results_tidy.html">results_tidy()</a></code> 
        </dt>
        <dd>Get benchmark results from `scbench` object in a tidy format, without
performance metrics</dd>
      </dl><dl><dt>
          
          <code><a href="plt_comp_performance.html">plt_comp_performance(<i>&lt;scbench&gt;</i>)</a></code> <code><a href="plt_comp_performance.html">plt_comp_performance(<i>&lt;hscreference&gt;</i>)</a></code> <code><a href="plt_comp_performance.html">plt_comp_performance(<i>&lt;screference&gt;</i>)</a></code> 
        </dt>
        <dd>Plot computational metrics computed during deconvolution</dd>
      </dl><dl><dt>
          
          <code><a href="plt_cor_heatmap.html">plt_cor_heatmap()</a></code> 
        </dt>
        <dd>Plot heatmap of correlation between deconvolution results for simulated
populations and true mixtures used for the pseudobulks</dd>
      </dl><dl><dt>
          
          <code><a href="plt_cors_scatter.html">plt_cors_scatter()</a></code> 
        </dt>
        <dd>Plot correlations by population between deconvolution results
for simulated populations and true mixtures used for the pseudobulk</dd>
      </dl><dl><dt>
          
          <code><a href="plt_lod_heatmap.html">plt_lod_heatmap()</a></code> 
        </dt>
        <dd>Plot heatmap summarizing limit of detection by population and method</dd>
      </dl><dl><dt>
          
          <code><a href="plt_lod_scatter.html">plt_lod_scatter()</a></code> 
        </dt>
        <dd>Plot limit of detection of deconvolution for each population as the proportion
of the pseudobulk where the method estimation is higher than at proportion=0 for
each population</dd>
      </dl><dl><dt>
          
          <code><a href="plt_population_mixtures.html">plt_population_mixtures()</a></code> 
        </dt>
        <dd>Plot sample population mixtures from an `scbench` object</dd>
      </dl><dl><dt>
          
          <code><a href="plt_rmse_heatmap.html">plt_rmse_heatmap()</a></code> 
        </dt>
        <dd>Plot RMSE between deconvolution results for simulated populations and
true mixtures used for the pseudobulks</dd>
      </dl><dl><dt>
          
          <code><a href="plt_spillover_heatmap.html">plt_spillover_heatmap()</a></code> 
        </dt>
        <dd>Plot heatmap summarizing spillover RMSE by population and method</dd>
      </dl><dl><dt>
          
          <code><a href="plt_spillover_scatter.html">plt_spillover_scatter()</a></code> 
        </dt>
        <dd>Plot correlations between deconvolution results and spillover mixtures
between each pair of populations</dd>
      </dl></div><div class="section level2">
      <h2 id="extras">Extras<a class="anchor" aria-label="anchor" href="#extras"></a></h2>
      
      <p class="section-desc"></p><p>Other package functions</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="read_h5ad.html">read_h5ad()</a></code> 
        </dt>
        <dd>Read a h5ad file and convert it to a Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="test_markers.html">test_markers()</a></code> 
        </dt>
        <dd>Evaluate annotation labels using Seurat markers</dd>
      </dl><dl><dt>
          
          <code><a href="install_cibersortx.html">install_cibersortx()</a></code> 
        </dt>
        <dd>Installs CIBERSORTx container using docker or singularity</dd>
      </dl><dl><dt>
          
          <code><a href="dot-pp_PBMCs.html">.pp_PBMCs()</a></code> 
        </dt>
        <dd>Get Seurat PBMCs pre-processed example</dd>
      </dl></div><div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>
      
      <p class="section-desc"></p><p>Included data</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="pbmc.html">pbmc</a></code> 
        </dt>
        <dd>Seurat example PBCM dataset</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Clarice S Groeneveld.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

